cpp/src/sampling/uniform_neighbor_sampling_impl.hpp

```diff
@@ -0,0 +1,191 @@
+/*
+ * Copyright (c) 2022, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <sampling/detail/graph_functions.hpp>
+
+#include <cugraph/detail/shuffle_wrappers.hpp>
+#include <cugraph/detail/utility_wrappers.hpp>
+#include <cugraph/graph.hpp>
+#include <cugraph/partition_manager.hpp>
+
+#include <raft/handle.hpp>
+
+#include <rmm/device_uvector.hpp>
+
+#include <cugraph-ops/graph/sampling.hpp>
+
+#include <algorithm>
+#include <limits>
+#include <numeric>
+#include <type_traits>
+#include <vector>
+
+namespace cugraph {
+namespace detail {
+
+template <typename graph_view_t>
+std::tuple<rmm::device_uvector<typename graph_view_t::vertex_type>,
+           rmm::device_uvector<typename graph_view_t::vertex_type>,
+           rmm::device_uvector<typename graph_view_t::weight_type>>
+uniform_nbr_sample_impl(
+  raft::handle_t const& handle,
+  graph_view_t const& graph_view,
+  rmm::device_uvector<typename graph_view_t::vertex_type>& d_in,
+  raft::host_span<const int> h_fan_out,
+  rmm::device_uvector<typename graph_view_t::edge_type> const& global_out_degrees,
+  rmm::device_uvector<typename graph_view_t::edge_type> const& global_degree_offsets,
+  rmm::device_uvector<typename graph_view_t::edge_type> const& global_adjacency_list_offsets,
+  bool with_replacement,
+  uint64_t seed)
+{
+  using vertex_t = typename graph_view_t::vertex_type;
+  using edge_t   = typename graph_view_t::edge_type;
+  using weight_t = typename graph_view_t::weight_type;
+
+  namespace cugraph_ops = cugraph::ops::gnn::graph;
+
+  CUGRAPH_EXPECTS(h_fan_out.size() > 0,
+                  "Invalid input argument: number of levels must be non-zero.");
+
+  rmm::device_uvector<vertex_t> d_result_src(0, handle.get_stream());
+  rmm::device_uvector<vertex_t> d_result_dst(0, handle.get_stream());
+  auto d_result_indices =
+    thrust::make_optional(rmm::device_uvector<weight_t>(0, handle.get_stream()));
+
+  size_t level{0};
+  size_t num_rows{1};
```

Comment:
Better re-name this to `row_comm_size` (this is more of a consistency thing).

Reply:
Done